### PR TITLE
Add notifier go client

### PR DESCRIPTION
**Pull request by ssd04**
Move notifier client form `notifier-go` repo to `elrond-go`, in order to avoid importing notifier component, which is not being used in elrond-go.
Notifier go PR: https://github.com/ElrondNetwork/notifier-go/pull/43

### CODE REVIEW

File: outport/factory/notifierFactory.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+	"github.com/ElrondNetwork/elrond-go-core/core"
+	"github.com/ElrondNetwork/elrond-go-core/core/check"
+	"github.com/ElrondNetwork/elrond-go-core/hashing"
+	"github.com/ElrondNetwork/elrond-go-core/marshal"
+	"github.com/ElrondNetwork/elrond-go/outport"
+	"github.com/ElrondNetwork/elrond-go/outport/notifier"
+)
+
+// ErrNilPubKeyConverter signals that a nil pubkey converter has been provided
+var ErrNilPubKeyConverter = errors.New("nil pub key converter")
```
Comment (bogdan-rosianu) [inline outport/factory/notifierFactory.go:15]: could have moved to an `errors.go` file, or unexport it

File: go.mod [hunk 2]
```diff
@@ -24,6 +23,7 @@ require (
 	github.com/gin-gonic/gin v1.7.7
 	github.com/gizak/termui/v3 v3.1.0
 	github.com/gogo/protobuf v1.3.2
+	github.com/google/go-cmp v0.5.6 // indirect
```
Comment (bogdan-rosianu) [inline go.mod:26]: might run `go mod tidy` again ? also, you could remove `go.sum` and run `go mod tidy` again, maybe these `// indirect` will disappear

File: outport/factory/notifierFactory_test.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+	t.Run("nil pub key converter", func(t *testing.T) {
+		t.Parallel()
+
+		args := createMockNotifierFactoryArgs()
+		args.PubKeyConverter = nil
+
+		en, err := factory.CreateEventNotifier(args)
+		require.Nil(t, en)
+		require.Equal(t, factory.ErrNilPubKeyConverter, err)
+	})
```
Comment (bogdan-rosianu) [inline outport/factory/notifierFactory_test.go:60]: could have added an unit test for the successful flow, as well

File: outport/notifier/eventNotifier.go [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+		httpClient:      args.HttpClient,
+		marshalizer:     args.Marshalizer,
+		hasher:          args.Hasher,
+		pubKeyConverter: args.PubKeyConverter,
+	}, nil
+}
+
+// SaveBlock converts block data in order to be pushed to subscribers
+func (en *eventNotifier) SaveBlock(args *indexer.ArgsSaveBlockData) error {
+	log.Debug("SaveBlock called at block", "block hash", args.HeaderHash)
```
Comment (bogdan-rosianu) [inline outport/notifier/eventNotifier.go:81]: might add more context on these logs, for example `eventNotifier: SaveBlock called...`

File: outport/notifier/eventNotifier.go [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+func (en *eventNotifier) SaveAccounts(_ uint64, _ []nodeData.UserAccountHandler) error {
+	return nil
+}
+
+// IsInterfaceNil returns whether the interface is nil
+func (en *eventNotifier) IsInterfaceNil() bool {
+	return en == nil
+}
+
+func (en *eventNotifier) Close() error {
```
Comment (bogdan-rosianu) [inline outport/notifier/eventNotifier.go:208]: missing comment

File: outport/notifier/httpClient.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+	if h.useAuthorization {
+		req.SetBasicAuth(h.username, h.password)
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		bodyCloseErr := resp.Body.Close()
```
Comment (bogdan-rosianu) [inline outport/notifier/httpClient.go:74]: might worth checking `if resp != nil && resp.Body != nil`

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

File: outport/factory/notifierFactory.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+// EventNotifierFactoryArgs defines the args needed for event notifier creation
+type EventNotifierFactoryArgs struct {
+	Enabled          bool
+	UseAuthorization bool
+	ProxyUrl         string
+	Username         string
+	Password         string
+	Marshalizer      marshal.Marshalizer
```
Comment (iulianpascalau) [inline outport/factory/notifierFactory.go:24]: can we rename the field Marshaller?

File: outport/factory/notifierFactory_test.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+		Marshalizer:      &testscommon.MarshalizerMock{},
+		Hasher:           &hashingMocks.HasherMock{},
+		PubKeyConverter:  &testscommon.PubkeyConverterMock{},
+	}
+}
+
+func TestCreateEventNotifier(t *testing.T) {
+	t.Parallel()
+
+	t.Run("nil marshalizer", func(t *testing.T) {
```
Comment (iulianpascalau) [inline outport/factory/notifierFactory_test.go:29]: nil marshaller?

File: outport/notifier/eventNotifier.go [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+}
+
+// SaveRoundsInfo returns nil
+func (en *eventNotifier) SaveRoundsInfo(_ []*indexer.RoundInfo) error {
+	return nil
+}
+
+// SaveValidatorsRating returns nil
+func (en *eventNotifier) SaveValidatorsRating(_ string, _ []*indexer.ValidatorRatingInfo) error {
+	return nil
```
Comment (iulianpascalau) [inline outport/notifier/eventNotifier.go:190]: might add tests for all these "mock" functions as to be sure the code does not panic when called

File: outport/notifier/httpClient.go [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+// HttpClientArgs defines the arguments needed for http client creation
+type HttpClientArgs struct {
+	UseAuthorization bool
+	Username         string
+	Password         string
+	BaseUrl          string
+}
+
+// NewHttpClient creates an instance of httpClient which is a wrapper for http.Client
+func NewHttpClient(args HttpClientArgs) *httpClient {
```
Comment (iulianpascalau) [inline outport/notifier/httpClient.go:36]: unit tests using a test http server?

File: outport/notifier/eventNotifier.go (hunk 4 above)
Comment (ssd04) [reply outport/notifier/eventNotifier.go]: added eventNotifier in log message

File: outport/notifier/httpClient.go (hunk 6 above)
Comment (ssd04) [reply outport/notifier/httpClient.go]: added additional check

File: outport/factory/notifierFactory.go (hunk 7 above)
Comment (ssd04) [reply outport/factory/notifierFactory.go]: :+1: renamed

File: outport/notifier/eventNotifier.go (hunk 9 above)
Comment (ssd04) [reply outport/notifier/eventNotifier.go:190]: added more tests

File: outport/notifier/eventNotifier_test.go [hunk 11]
```diff
@@ -111,3 +111,26 @@ func TestFinalizedBlock(t *testing.T) {
 
 	require.True(t, wasCalled)
 }
+
+func TestMockFunctions(t *testing.T) {
+	t.Parallel()
+
```
Comment (iulianpascalau) [inline outport/notifier/eventNotifier_test.go:117]: could have added:
```GO
    defer func(){
        r := recover()
        if r != nil {
            assert.Fail(t, fmt.Sprintf("should have not panicked: %v", r))
        }
    }()
```

File: outport/notifier/httpClient_test.go [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+	testPayload := testStruct{
+		Hash: "hash1",
+	}
+
+	err := client.Post("/events/push", testPayload, nil)
+	require.Nil(t, err)
+
+	require.True(t, wasCalled)
+}
```
Comment (iulianpascalau) [inline outport/notifier/httpClient_test.go:60]: from linter:
factory/statusComponents.go
L250: 
```
unknown field 'Marshalizer' in struct literal of type "github.com/ElrondNetwork/elrond-go/outport/factory".EventNotifierFactoryArgs
```

**Review by iulianpascalau — DISMISSED** [review]

File: outport/notifier/eventNotifier_test.go (hunk 11 above)
Comment (ssd04) [reply outport/notifier/eventNotifier_test.go:119]: :+1: added recover check

**Review by iulianpascalau — DISMISSED** [review]

**Review by bogdan-rosianu — DISMISSED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by gabi-vuls — APPROVED** [review]
@@                            Log scanner                                  @@
##                            add-notifier-go-client                       ##
================================================================================
- Known Warnings              12
- New Warnings                2
- Known Errors                0
- New Errors                  0
- Panics                      0
================================================================================

**Pull request merged**